### PR TITLE
Joss paper

### PR DESCRIPTION
**Pull request by ryanmrichard**
This PR contains the JOSS manuscript (well the source code for it). The JOSS GitHub action will turn that source code into the actual manuscript.

We have no intention of merging this branch with master, rather this PR serves as a central location for collaborating on the manuscript.

For reference:
- [this](https://joss.readthedocs.io/en/latest/submitting.html#what-should-my-paper-contain) is what JOSS recommends putting in the paper. 
- Papers need to be 500 to 1000 words (we're at like 900).

### CODE REVIEW

File: paper.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
+---
+title: 'CMinx: Restructured Text Documentation Generator for CMake'
```
Comment (AutonomicPerfectionist) [inline paper.md:2]: I would say the most important part of the title is that it's for CMake, so it would probably be best if we could reword it to move CMake closer to the beginning of the title. Maybe "CMinx: The CMake Documentation Generator"? Or "CMinx: Documenting CMake with Restructured Text"?

With the current title, readers might miss the CMake part and assume it's an alternative to already existing RST documentation generators

File: paper.bib [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    url = {https://stackoverflow.com/a/54671996}
+}
+
+@misc{sphinx_plugin,
+    author = {Kitware},
+    title = {Sphinx Domain for Modern CMake},
+    year = {2022},
+    publisher = {GitHub},
+    journal = {GitHub Repository},
+    url = {https://github.com/scikit-build/moderncmakedomain}
```
Comment (AutonomicPerfectionist) [inline paper.bib:25]: We should probably look into integrating with this domain in the future as an optional extension since it seems a lot of "official" CMake documentation uses it

File: paper.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+to the documentation, listing the function's parameters, and the overall
+formatting (*e.g.*, titles and subtitles, creation of parameter 
+tables, underline/overline consistency). For build system developers 
+maintaining one or two CMake modules, these are admittedly pretty minor 
+inconveniences; however, for an organizations which maintaining a substantial 
+CMake code base, these "minor inconveniences" can impact productivity. 
+In order to increase the productivity of the CMakePP team, we created CMinx.
+
+Concurrent with the submission of this manuscript, we have also released the 
+first production version of CMinx. Promisingly, the CMinx GitHub organization
```
Comment (AutonomicPerfectionist) [inline paper.md:114]: > first production version of CMinx

This would be a great place to put a reference to the PyPI page. We should probably also mention that it has a CMake interface somewhere and put a link to the GitHub page there

**Review by AutonomicPerfectionist — CHANGES_REQUESTED** [review]
@ryanmrichard I read through the paper and marked a few changes I would suggest. I think right now we cover pretty much all the bases, if we really wanted to we could include something about support for CMakePP/CMakeTest stuff but that would probably be outside the scope of the paper in my opinion.

File: paper.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+software development ecosystem and it is our hope that CMinx will serve as the
+foundation for a robust CMake development ecosystem.
+
+Anecdotal evidence [@official_solution] indicates that the developers of the
+CMake language (*i.e.*, Kitware) internally write their API documentation
+using reST and Sphinx. Following best practices, this reST API documentation
+resides next to the CMake code being described. Kitware has also written a
+Sphinx plugin that makes it easy to extract the API documentation as part of a
+normal Sphinx workflow. This Sphinx plugin is distributed with the source code
+for the CMake interpreter, and is also available in a GitHub repository mirror
```
Comment (AutonomicPerfectionist) [inline paper.md:98]: > This Sphinx plugin is distributed with the source code
for the CMake interpreter,

Comma after "interpreter" should be removed

File: paper.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+[@sphinx_plugin]. For completeness, we note that similar Sphinx plugins
+[@official_sphinx_domain; @marco_koch] have been independently developed, but
```
Comment (AutonomicPerfectionist) [inline paper.md:100]: > have been independently developed, but

Comma after "developed" should be removed

File: paper.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+appear to now be abandoned.
+
+To our knowledge, all of the aforementioned Sphinx plugins simply extract the
+reST API documentation verbatim. Notably this means the developer is responsible
```
Comment (AutonomicPerfectionist) [inline paper.md:104]: > Notably this means the developer

Should be a comma after "notably"

File: paper.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+build system developers maintaining one or two CMake modules, these are
+admittedly pretty minor inconveniences; however, for organizations
+maintaining a substantial CMake code base, these "minor inconveniences" can
+impact productivity, particularly when ensuring consistency. CMinx
+differs from previous solutions primarily in three ways. First, CMinx
+understands the grammar of the CMake language, meaning CMinx can automatically
+generate some of the documentation by "reading" the source code. Second, CMinx
+generates static reST files; this decreases the number of stub files developers
+need to maintain and makes it easier for the resulting documentation to be
+used in workflows which do not rely on Sphinx. Finally, CMinx has a CMake API
```
Comment (AutonomicPerfectionist) [inline paper.md:117]: > used in workflows which do not rely on Sphinx.

"Which" should be swapped for "that"

Comment (AutonomicPerfectionist) [inline paper.md:117]: > Finally, CMinx has a CMake API
so that it integrates more easily into existing CMake workflows.

I'm not quite sure why but something doesn't seem quite right about this sentence. Maybe something like "Finally, CMinx has a CMake API to integrate more easily with existing CMake workflows"?

Or perhaps something like "Finally, CMinx has a CMake API to allow easy integration with existing CMake workflows"

**Review by AutonomicPerfectionist — CHANGES_REQUESTED** [review]
@ryanmrichard most of it looks pretty good now, just noticed mainly some comma issues here and there. Also, for whatever reason the PR is now listing several source files as being changed, looks to be the stuff I did in #100. Probably not an issue though since this PR isn't going to be merged

**Comment by ryanmrichard** [comment]
@AutonomicPerfectionist it's submitted :tada:

File: paper.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+character). The documentation comments, written in reST [@rest],
+describe to the reader how the functions, parameters, and variables should be
+used. Running CMinx on the annotated source code generates reST files containing
+the API documentation. The reST files can then be converted into static
+websites with tools such as Sphinx [@sphinx] or
+easily converted to another format via Pandoc [@pandoc].
+
+Unlike other solutions for documenting CMake modules, CMinx knows the CMake
+language's grammar. This enables CMinx to automatically extract function/macro
+signatures (even when functions are not documented). CMinx also integrates
```
Comment (peanutfun) [inline paper.md:43]: Avoid parentheses
```suggestion
signatures, even when functions are not documented. CMinx also integrates
```

File: paper.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+seamlessly into existing CMake build systems. CMinx's output is highly
+customizable and easily controlled via a YAML [@yaml] configuration file.
```
Comment (peanutfun) [inline paper.md:45]: I can find this claim nowhere in the documentation. How can CMinx be customized with a configuration file? Please clarify or remove the claim.

File: paper.md [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+The exascale-era of high-performance scientific computing arguably started when
+the Frontier supercomputer achieved a performance of more than 1.1 exaFLOPs
+(10$^{18}$ floating-point operations per second) on the High-Performance
+Linpack Benchmark [@frontier]. With the breakdown of Dennard scaling
+(*i.e.*, the power usage of a transistor is proportional to its area),
+achieving exascale performance is presently only possible by relying
+on heterogeneous hardware. Unfortunately, hardware heterogeneity significantly
+complicates the process of developing high-performance scientific software.
+For many high-performance software packages, these complications start in the
+build system.
```
Comment (peanutfun) [inline paper.md:62]: I find that this paragraph to be superfluous. The issue of maintaining build system code has little to do with exascale computing per se. The actual connection is done in the following paragraph, so I would suggest to remove this paragraph entirely. I think the paper actually works well without it.


```suggestion
```

File: paper.md [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+The process of building a software package written in a compiled language
+(*e.g.*, C or C++) has always been complicated. Over the years, various build
+system solutions have evolved to ease the process. Historically, there
+has been a propensity to treat each build system as a one-off use case. This is
+understandable since build systems have tended to be relatively small (*i.e.*,
+less than ~1K lines of code) and tightly coupled to the identity of the package.
```
Comment (peanutfun) [inline paper.md:69]: Unnecessary specification in parentheses. However, "identity" is too vague in my opinion:
```suggestion
understandable since build systems have tended to be relatively small
and tightly coupled to the structure and purpose of the package.
```

File: paper.md [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+With build system complexity at an all-time high, there is now an
+increasing desire (at least within scientific computing communities) to treat
+the underlying build system infrastructure as code. This means that the
+build system should be modularized, and those modules should be documented,
+tested, and reusable. With the popularity of C/C++ for high-performance
```
Comment (peanutfun) [inline paper.md:74]: I guess many developers will agree, but evidence appears to be anecdotal. I think this section (and the entire paragraph) would be improved with some references on the topic, although I realize they are hard to come by. I found an (probably outdated?) article on the state of exascale computing in the US (https://doi.org/10.1177%2F1094342014522573) and a paper on the xSDK foundations (https://arxiv.org/pdf/1702.08425.pdf), which both identify (meta-)build systems as essential cornerstones for modern HPC applications. Maybe you could reference them or similar publications?

File: paper.md [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+computing, "build system" is increasingly becoming synonymous with CMake
+(*N.B.* CMake is often described as a build system generator; however, since
+the typical CMake workflow encapsulates running the build system, the
+present manuscript ignores the "generator" distinction). Therefore with the
```
Comment (peanutfun) [inline paper.md:78]: Too specific. This could be moved into a footnote or left out entirely.

File: paper.md [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+can be created by writing CMake modules. While the CMake language is
+flexible and relatively simple, it is not without its pitfalls. Unfortunately,
+tools  to facilitate the development of CMake modules are relatively sparse.
+Here we introduce CMinx, a tool for generating API documentation for CMake
+modules. API documentation is arguably one of the most basic elements of a
+software development ecosystem, and it is our hope that CMinx will serve as the
+foundation for a robust CMake development ecosystem.
+
+Anecdotal evidence [@official_solution] indicates that the developers of the
+CMake language (*i.e.*, Kitware) internally write their API documentation
```
Comment (peanutfun) [inline paper.md:95]: Avoid parentheses
```suggestion
Anecdotal evidence [@official_solution] indicates that Kitware, the developers
of the CMake language, internally write their API documentation
```

File: paper.md [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+generate some of the documentation by "reading" the source code. Second, CMinx
+generates static reST files; this decreases the number of stub files developers
+need to maintain and makes it easier for the resulting documentation to be
+used in workflows that do not rely on Sphinx. Finally, CMinx has a CMake API
+to integrate more easily into existing CMake workflows.
+
+Concurrent with the submission of this manuscript, we have also released the
+first production version of CMinx, version 1.0.0. CMinx can be obtained
+from the Python Packaging Index via "`pip install CMinx`". Alternatively, CMinx
+can be used as a CMake module via CMake's "`FetchContent`" command.
```
Comment (peanutfun) [inline paper.md:127]: Could you add an example for the `FetchContent` solution to the documentation? I think this would be very helpful, as the CMake modules would become available to users.

File: paper.md [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+[@sphinx_plugin]. For completeness, we note that similar Sphinx plugins
+[@official_sphinx_domain; @marco_koch] have been independently developed but
+appear to now be abandoned.
+
+To our knowledge, all of the aforementioned Sphinx plugins simply extract the
+reST API documentation verbatim. Notably, this means the developer is
+responsible
+for manually: adding the function/macro signatures to the documentation, listing
+the function's parameters, and the overall formatting (*e.g.*, titles and
+subtitles, creation of parameter tables, underline/overline consistency). For
```
Comment (peanutfun) [inline paper.md:110]: Again, the examples are specific in my perspective. Also, the colon appears a bit stray to me.
```suggestion
for manually adding the function/macro signatures to the documentation, listing
the function's parameters, and the overall formatting. For
```

File: paper.bib [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    journal = {Stack Overflow Forum},
+    url = {https://stackoverflow.com/a/54671996}
+}
+
+@misc{sphinx_plugin,
+    author = {Kitware},
+    title = {Sphinx Domain for Modern CMake},
+    year = {2022},
+    publisher = {GitHub},
+    journal = {GitHub Repository},
```
Comment (peanutfun) [inline paper.bib:24]: I think this renders badly in the resulting paper, although the JOSS example paper does it like that. I would propose to use
```suggestion
    howpublished = {{GitHub Repository}},
```

File: paper.bib [hunk 18]
```diff
@@ -0,0 +1,77 @@
+@misc{frontier,
+    author = {Barb Helland},
+    title = {At the Frontier: DOE Supercomputing Launches the Exascale Era},
+    year = {2022},
+    publisher = {United States Department of Energy Office of Science},
+    journal = {Website},
+    url = {https://www.energy.gov/science/articles/frontier-doe-supercomputing-launches-exascale-era}
+}
```
Comment (peanutfun) [inline paper.bib:8]: I think this can be considered an `article`, but anyways, "Website" is not a journal.
```suggestion
@article{frontier,
    author = {Barb Helland},
    title = {At the Frontier: DOE Supercomputing Launches the Exascale Era},
    year = {2022},
    publisher = {United States Department of Energy Office of Science},
    url = {https://www.energy.gov/science/articles/frontier-doe-supercomputing-launches-exascale-era}
}
```

File: paper.md [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+
+Concurrent with the submission of this manuscript, we have also released the
+first production version of CMinx, version 1.0.0. CMinx can be obtained
+from the Python Packaging Index via "`pip install CMinx`". Alternatively, CMinx
+can be used as a CMake module via CMake's "`FetchContent`" command.
+Despite only just releasing 1.0.0, the CMinx GitHub organization
+has already started to see attention and interest from developers not
+affiliated with the authors. So while CMinx might not be considered research
+software, we anticipate CMinx to be a useful productivity tool for the large
+swath of research software that uses CMake as their build system.
```
Comment (peanutfun) [inline paper.md:132]: This entire paragraph does not fit into the "Statement of need" and could instead be moved to the end of the "Summary", in my opinion.

**Review by peanutfun — CHANGES_REQUESTED** [review]
I thought I would add my suggestions and comments on the paper right in this PR. Generally, I think it is very well written and gives a nice overview and statement of need for CMinx. Please note that all suggestions are to be treated as such. Feel free to discuss, alter and reject them.

My suggestions are mostly related to the profuse use of parentheses, which make many sentences hard to read. Many of them can simply be integrated into the sentence, others contain too specific information and could be dropped entirely. I also think that the paragraph on exascale computing is off-topic and can be dropped without impairing the rest of the paper. Finally, I would be happy to see some references to the importance of build system maintenance.

File: paper.md (hunk 9 above)
Comment (ryanmrichard) [reply paper.md:45]: This is a relatively new feature. At the moment I think it's only documented in the ``cminx --help`` message. I'll make sure that gets into the documentation.

File: paper.md [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+# Summary
+
+This manuscript introduces CMinx, a program for generating
+application programming interface (API) documentation for CMake modules
+(code written in the CMake language). Since most of CMinx's intended audience
+is comprised of C/C++ developers, CMinx is designed to operate similar to
+Doxygen [@doxygen] (the *de facto* C/C++ API documentation tool).
+Specifically, developers annotate their CMake source with documentation
+comments (traditional CMake block comments, starting with an extra "`[`"
+character). The documentation comments, written in reST [@rest],
```
[the remark this answers is not in the document]
Comment (ryanmrichard) [reply paper.md:34]: I'm inclined to keep this parenthetical since I think it's important to introduce the "documentation comment" term, and define it.

[the remark this answers is not in the document]
Comment (peanutfun) [reply paper.md:34]: Parentheses can state something less important than the parent sentence, something that can be skipped. If you want to define the term here, I suggest to make it more clear. How about

> Specifically, developers annotate their CMake source with "documentation comments", which are traditional CMake block comments starting with an extra "`[`" character.

File: paper.md (hunk 19 above)
Comment (ryanmrichard) [reply paper.md]: Done. I moved it to the end of the summary section.

File: paper.md (hunk 13 above)
Comment (ryanmrichard) [reply paper.md]: I dropped it.

File: paper.md (hunk 12 above)
Comment (ryanmrichard) [reply paper.md]: I was only aware of anecdotal evidence. So I greatly appreciate you finding some actual literature references. I went a head and added those references.

File: paper.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (ryanmrichard) [reply paper.md]: I like your suggestion.

File: paper.md [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+Despite only just releasing 1.0.0, the CMinx GitHub organization
+has already started to see attention and interest from developers not
+affiliated with the authors. So while CMinx might not be considered research
+software, we anticipate CMinx to be a useful productivity tool for the large
+swath of research software that uses CMake as their build system.
+
+# Statement of need
+
+The process of building a software package written in a compiled language
+(*e.g.*, C or C++) has always been complicated. Over the years, various build
```
Comment (robertodr) [inline paper.md:64]:
```suggestion
(*e.g.*, C, C++ or Fortran) has always been complicated. Over the years, various build
```
I'd add Fortran as well, as that adds its own intricacies to the build system.

File: paper.md [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+system solutions have evolved to ease the process. Historically, there
+has been a propensity to treat each build system as a one-off use case. This is
+understandable since build systems have tended to be relatively small
+and tightly coupled to the structure and purpose of the package.
+With build system complexity at an all-time high[@snir2014,@xSDK], there is an
```
Comment (robertodr) [inline paper.md:69]:
```suggestion
With build system complexity at an all-time high [@snir2014,@xSDK], there is an
```
should there be a space here?

File: paper.md [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+increasing need to treat the underlying build system infrastructure as code.
+This means that the build system should be modularized, and those modules
+should be documented, tested, and reusable. With the popularity of C/C++ for
+high-performance computing, "build system" is increasingly becoming synonymous
+with CMake. Therefore with the advent of the exascale era, there is a desperate
```
Comment (robertodr) [inline paper.md:74]:
```suggestion
with CMake: there is a desperate
```
I think the need is there whether you're application targets the exascale or not.

File: paper.bib [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+              Al Geist and Rinku Gupta and Fred Johnson and
+              Sriram Krishnamoorthy and Sven Leyffer and Dean Liberty and
+              Subhasish Mitra and Todd Munson and Rob Schreiber and
+              Jon Stearley and Eric Van Hensbergen},
+  journal  = {The International Journal of High Performance Computing
+              Applications},
+  title    = {Addressing failures in exascale computing},
+  year     = {2014},
+  pages    = {129-173},
+  volume   = {28},
```
Comment (peanutfun) [inline paper.bib:106]: Add DOI
```suggestion
  volume   = {28},
  doi = {10.1177/1094342014522573},
```

File: paper.md [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+
+# Statement of need
+
+The process of building a software package written in a compiled language
+(*e.g.*, C, C++ or Fortran) has always been complicated. Over the years, various build
+system solutions have evolved to ease the process. Historically, there
+has been a propensity to treat each build system as a one-off use case. This is
+understandable since build systems have tended to be relatively small
+and tightly coupled to the structure and purpose of the package.
+With build system complexity at an all-time high [@snir2014,@xSDK], there is an
```
Comment (peanutfun) [inline paper.md:69]: Needs semicolon to render properly
```suggestion
With build system complexity at an all-time high [@snir2014; @xSDK], there is an
```

File: paper.bib [hunk 26]
```diff
@@ -0,0 +1,107 @@
+@article{frontier,
+    author = {Barb Helland},
+    title = {At the Frontier: DOE Supercomputing Launches the Exascale Era},
+    year = {2022},
+    publisher = {United States Department of Energy Office of Science},
+    url = {https://www.energy.gov/science/articles/frontier-doe-supercomputing-launches-exascale-era}
+}
+
```
Comment (peanutfun) [inline paper.bib:8]: This is not referenced anymore, as the respective paragraph was deleted. I'm not sure if it fits into the new structure. If not, the entire entry can be removed.
```suggestion
```

File: paper.bib [hunk 27]
```diff
@@ -0,0 +1,107 @@
...
+@misc{pandoc,
+    title ={Pandoc: A universal document converter},
+    url = {https://pandoc.org/},
+    year = {2022}
+}
+
+@misc{xSDK,
+  url = {https://arxiv.org/abs/1702.08425},
```
Comment (peanutfun) [inline paper.bib:78]: DOI might be safer than URL
```suggestion
  doi = {10.48550/arXiv.1702.08425},
```

**Review by peanutfun — CHANGES_REQUESTED** [review]
I found some minor issues with the latest version of the paper. They are very easy to update: I suggest adding DOIs for the articles, fixing one of the citations in the text which does not render due to a syntax error, and removing an unused reference (or alternatively citing it in the text where appropriate).

**Comment by ryanmrichard** [comment]
@peanutfun thanks for your careful reviewing! I went ahead and made those changes.

**Comment by ryanmrichard** [comment]
Now that the paper is published I'm going to delete this branch and close this PR. If people want to read the paper they can follow the link in the little DOI button.
